zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/metrics/UsageMetricsExportProcessor.java

```diff
@@ -32,95 +28,64 @@ public class UsageMetricsExportProcessor implements TypedRecordProcessor<UsageMe
 
   private static final Logger LOG = LoggerFactory.getLogger(UsageMetricsExportProcessor.class);
 
-  private final MutableUsageMetricState usageMetricState;
+  private final UsageMetricState usageMetricState;
   private final StateWriter stateWriter;
   private final KeyGenerator keyGenerator;
+  private final InstantSource clock;
 
   public UsageMetricsExportProcessor(
-      final MutableUsageMetricState usageMetricState,
+      final UsageMetricState usageMetricState,
       final Writers writers,
-      final KeyGenerator keyGenerator) {
+      final KeyGenerator keyGenerator,
+      final InstantSource clock) {
     this.usageMetricState = usageMetricState;
     stateWriter = writers.state();
     this.keyGenerator = keyGenerator;
+    this.clock = clock;
   }
 
   @Override
   public void processRecord(final TypedRecord<UsageMetricRecord> usageMetricRecord) {
 
-    final UsageMetricRecord eventRecord =
-        new UsageMetricRecord()
-            .setIntervalType(IntervalType.ACTIVE)
-            .setEventType(EventType.NONE)
-            .setResetTime(usageMetricRecord.getTimestamp());
-
-    final var bucket = usageMetricState.getActiveBucket();
-    if (bucket == null || !bucket.isInitialized()) {
-      appendFollowUpEvent(eventRecord);
-      return;
-    }
-
-    final var isRPIMapEmpty = bucket.getTenantRPIMap().isEmpty();
-    final var isEDIMapEmpty = bucket.getTenantEDIMap().isEmpty();
-    final var isTUMapEmpty = bucket.getTenantTUMap().isEmpty();
+    final long now = clock.millis();
 
-    final var events = new ArrayList<UsageMetricRecord>();
-    if (!isRPIMapEmpty || !isEDIMapEmpty || !isTUMapEmpty) {
-      processMetricType(
-              bucket, eventRecord, EventType.RPI, isRPIMapEmpty, bucket.getTenantRPIMapValue())
-          .ifPresent(events::add);
-      processMetricType(
-              bucket, eventRecord, EventType.EDI, isEDIMapEmpty, bucket.getTenantEDIMapValue())
-          .ifPresent(events::add);
-      processMetricType(
-              bucket, eventRecord, EventType.TU, isTUMapEmpty, bucket.getTenantTUMapValue())
-          .ifPresent(events::add);
-    } else {
-      events.add(eventRecord);
+    // bucket is initialized when some metric is recorded or when the first applier runs
+    var bucket = usageMetricState.getActiveBucket();
+    if (bucket == null) {
+      bucket = new PersistedUsageMetrics();
     }
 
-    // append events at the end so bucket is not reset while processing
-    events.forEach(this::appendFollowUpEvent);
+    // Copy and update the bucket fromTime as we are exporting it now.
+    // Additionally, ensure the toTime is set. This can happen if metrics were recorded before the
```

Comment:
🔧 I think you mixed from with to time here 😅

Reply:
Yepp 😄